api/pkg/transformer/pipeline/operation.go

```diff
@@ -1,7 +1,118 @@
 package pipeline
 
-import "context"
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/gojek/merlin/pkg/transformer/spec"
+	"github.com/gojek/merlin/pkg/transformer/types"
+	"github.com/gojek/merlin/pkg/transformer/types/table"
+)
 
 type Op interface {
 	Execute(context context.Context, environment *Environment) error
+	AddInputOutput(input, output map[string]interface{}) error
+	GetOperationTracingDetail() ([]types.TracingDetail, error)
+}
+
+type OperationIO struct {
+	Name  string
+	Value interface{}
+}
+
+type OperationTracing struct {
+	mutex  sync.RWMutex
+	Input  []map[string]interface{}
+	Output []map[string]interface{}
+	Specs  interface{}
+	OpType types.OperationType
+}
+
+func NewOperationTracing(operationSpecs interface{}, opType types.OperationType) *OperationTracing {
+	opTracing := &OperationTracing{
+		Specs:  operationSpecs,
+		OpType: opType,
+	}
+	return opTracing
+}
+
+func (ot *OperationTracing) AddInputOutput(input, output map[string]interface{}) error {
+	ot.mutex.Lock()
```

Comment:
Why is there need for mutex? Is there case that `AddInputOutput` is called concurrently?

Reply:
yeah, we do it sequentially. I removed the mutex